### ISSUE TITLE
[BUG] BlobClient.UploadAsync with an unseekable stream always buffers the entire stream

### ISSUE THREAD

**Issue by logiclrd** [Azure/azure-sdk-for-net#25011]
**Describe the bug**
When trying to write a lightweight service that receives content in a `Stream` and passes it on to a blob upload, I encountered pathological buffering of the entire `Stream` content.

I believe this to be caused by way the blob client implementation uses `HttpClient`.

* `BlobClient.UploadAsync(Stream)` calls the extension method `BlobOperationsExtensions.UploadAsync`
* `BlobOperationsExtensions.UploadAsync` calls `BlobOperations.UploadWithHttpMessagesAsync`
* `BlobOperations.UploadWithHttpMessagesAsync` sets up an `HttpRequestMessage`, setting its `Content` property to an instance of `StreamContent` with the supplied `Stream`
* This `HttpRequestMessage` is then delivered using `HttpClient.SendAsync`
* `HttpClient.SendAsync` handles the message using the `HttpClientHandler` class
* The `PrepareAndStartContentUpload` method in the `HttpClientHandler` class checks if the `HttpContent` object it has been given specifies a `Content-Length` header. **If no `Content-Length` header is provided, then it buffers the content in order to determine its length.**

But, the `HttpContent` in question is the `StreamContent` initialized earlier by `UploadWithHttpMessagesAsync` in `BlobOperations`, and it does not set a `Content-Length` header.

**Expected behavior**
Data is transferred from the stream in manageably-sized chunks, so that memory usage is well-known regardless of the stream size. Specifically, if the `Stream` is seekable, then the stream's `Length` should be used for the `Content-Length` header value automatically. (This strategy is used in other ways inside `StreamContent` in its `TryComputeLength` method.)

**Actual behavior (include Exception or Stack Trace)**
If the stream contains 500MB of data then the process memory usage will grow by 500MB before the blob upload operation actually initiates the network connection to the Azure servers. If the stream contains more than 2GB of data then the blob upload operation will fail because `MemoryStream` will refuse to store more than 2GB of data.

**To Reproduce**
Using a `BlobClient`, use an `UploadAsync` operation that accepts a `Stream` for the data. Provide a `FileStream` pointing at a large file for the effects on process memory to be more easily noticeable.

**Environment:**
 - Azure.Storage.Blobs 12.10.0
 - Windows 10 .NET Core 3.1
 - Visual Studio 16.11.3

**Comment by logiclrd** [comment]
My analysis, just to provide additional information, is based on two independent things:

1) Observing memory usage grow very rapidly in a process that is running on .NET Core 3.1 and uploading a large file.

2) Looking at the implementation of `UploadAsync` in GitHub.

I am poking at it further to gather more information and confirm or deny my hypothesis.

**Comment by logiclrd** [comment]
The first link, from `BlobClient.UploadAsync` to `BlobOperationsExtensions.UploadAsync`, is the result of GitHub's analysis and cross-linking, but on closer analysis, it's a false positive. There isn't actually any connection between those two things, as far as I can see.

But, I do see a massive increase in memory usage trying to upload a large file with `BlobClient.UploadAsync`.

**Comment by logiclrd** [comment]
I inspected the code and found that the upload process, in inherently supporting retries, requires that earlier data in the stream be accessible. As such, if you supply an unseekable stream, it _must_ copy it to a seekable buffer to support retries.

Further inspection revealed that it is possible to disable parallel transfers with `MaximumConcurrency` setting in `BlobUploadOptions` and set the size of the block it will treat as a unit in performing the upload with `MaximumTransferSize`.

This gave me an idea for a possible way to support upload of an unseekable stream in a resource-constrained environment: An intermediate stream with a ringbuffer of the most recently read data, allowing seeking but only as far back as the ringbuffer extends. This is not an entirely trivial thing to implement, but this morning I managed to create a working implementation.

With this implementation, I can wrap my unseekable stream in one of these `BufferWindowStream` objects, as I called them, with the ringbuffer size set to equal `MaximumTransferSize`. In the event of a failed block, it should only rewind the stream by the size of that one block, which will then re-read the data out of the ringbuffer before proceeding. (_I have not yet verified that it will never rewind by more than the block size._)

Running this code, I can upload a multi-GB file to Azure Blob Storage streaming it through an intermediate web server running on Kestrel without the web server's memory usage exceeding 90 MB.

Since the intersection of unseekable streams and very large amounts of data can't be _that_ rare, perhaps it would make sense to integrate this approach into `BlobClient` itself?

**Comment by logiclrd** [comment]
Update: I interrupted the network after the first block had been staged, which in theory should only have required retrying the second block, but the Azure Blob Storage SDK code wanted to rewind the stream right back to the start even though the first block should have already been sent. 🤔

**Comment by logiclrd** [comment]
Looking at the code more closely, in theory, letting it use `GetBufferedPartitionInternal` should be close to identical, functionally. But, there is no way to communicate the length of the stream if it is not seekable.

**Comment by logiclrd** [comment]
I believe I have resolved that issue. It required `StorageTransferOptions.InitialTransferSize` to be set to a value no larger than the ringbuffer size. :-)

So, I think I have a valid work-around, _but_ it's an awful lot of work to support large restartable uploads on non-seekable streams in a resource-constrained environment:

* It requires sophisticated code on my end to buffer only just as much of the stream as is needed to allow for retries.
* It requires `StorageTransferOptions` to be set up in just the right way.

I don't have an in-depth understanding of the code inside the Azure Blob Storage client, but if I understood what I saw correctly, it has the ability to do a partitioned upload in a way that limits memory allocation, but the current implementation simply cannot avoid full-stream buffering if the supplied `Stream` isn't seekable.

**Comment by amishra-dev** [comment]
@jaschrep-msft can we please create a documentation task to get this documented in the backlog?

**Comment by jaschrep-msft** [comment]
@logiclrd if your `StorageTransferOptions` are correct (which you seem to have figured out over the course of this issue) then this buffering as needed code is already in the SDK for you. You should be able to remove your intermediate buffer with your appropriately tuned transfer options and see a limitation on the amount of data you buffer at any given time.

- For safety/retry purposes, the SDK will not upload data from a stream that is not seekable
- We buffer unseekable data to make it seekable
- We buffer seekable data if parallelism is desired
- Due to the nature of the REST layer, entire REST uploads must be retried from the start
- `BlobClient` breaks up your large upload into multiple smaller REST uploads based on `StorageTransferOptions`
- Buffering of unseekable streams happens lazily on a per-REST basis
  1. a buffer of size n in acquired, where n is the size of the next REST upload to perform
  2. next n bytes are read from the unseekable stream into this buffer
  3. buffer is used as a seekable stream to perform a retryable upload
  4. buffer is released upon success
- Initial transfer size when dealing with an unseekable stream happens before any other uploads are attempted

Because of this, the SDK should never be buffering more than either `InitialTransferSize` or `MaximumTransferSize * MaximumConcurrency` on a rolling basis as it works through the provided unseekable stream.

I believe this bug to be nonexistent and only reported based on a misunderstanding of the API. I will be closing it. If there is concern that the above description of the BlobClient upload process is incorrect and we are actually buffering far more, feel free to reopen.

I will be scheduling work to make this information more discoverable, per @amishra-dev's request.

**Comment by logiclrd** [comment]
Indeed, it appears that the key thing was having `StorageTransferOptions` fully-populated. With default options, it would always try to ascertain the length before starting the upload, buffering the entire stream in the process. But, now that `InitialTransferSize` and `MaximumTransferSize` are set, the original unseekable stream can be provided without memory usage growing without bound.

I just did a transfer of a 2.7 GB file, and memory usage climbed continuously for a good 10% of the transfer, but more and more slowly as time went on, and eventually reversing direction and gradually dropping, slowly fluctuating.

* It started at 75 MB and climbed quickly at first.
* When the transfer reached 275 MB, memory usage was about 143 MB.
* By 425 MB, it was at 154 MB, and it appeared to have _essentially_ plateaued. Memory usage continued to fluctuate, for instance hitting about 162 MB around the 800 MB point, but it subsequently dropped back down as low as 140 MB.
* Just before the 2 GB mark, however, it had dropped back as far as 120 MB. By 2 GB it was back up to 132 MB.

I repeated the test with `MaximumConcurrency` set to 5 with very similar results. In both tests, I interrupted the network to verify it could restart and found no issues.

To be clear, forwarding this file transfer from an incoming web request to an outbound `BlobClient.UploadAsync` call is the _only_ thing the process was doing.

I'm not sure why memory usage increased so much with a single upload thread and a 512 KB buffer size. Perhaps it is something within the ASP.NET Core stack or Kestrel. In any case, it didn't end up trying to store the entire 2.7 GB file in a `MemoryStream` as it had originally.

Documentation that calls out the significance of the stream being seekable and the importance of `InitialTransferSize` and `MaximumTransferSize` for non-seekable streams would indeed be very helpful. :-)
